### DISCUSSION TITLE
Access the object graph from config.py

### DISCUSSION THREAD

**Discussion by ervinpopescu** [qtile/qtile#3455] [Q&A] [answered]
How does one do this exactly?

I've RTFM about the (Interactive)CommandClient, but that only works from outside the config files.

```log
asyncio.run() cannot be called from a running event loop
```

This makes sense, I can't call the client from inside another client(I hope I got this right).
I need to run a hook and that doesn't work with my current implementation outside of the config territory.
So:
a) Please help me figure out hooks and running them outside of the config 
OR
b) Please help me access the graph from the config

**Comment by elParaguayo** [comment, marked as answer]
You can't hook from outside the config and you don't *really* use the object graph from the config either (except in lazy calls). However, you can still do everything you want in the config. 

Can you explain what you're trying to do and we'll help show the best way to do it.

**Reply by ervinpopescu** [reply]
I'm trying to get a notification daemon working as much as a DE one as possible(group changes, window changes etc).
I remembered I "borrowed" some code from someone and saw there I can use `qtile` for my needs.
I should look at libqtile.core.manager.Qtile, right?
`qtile.current_screen.group` etc?

**Reply by elParaguayo** [reply]
Well, there's a notification server already within qtile (widget.Notify).

There's also a function in libqtile.utils called send_notification. Using that with the hooks should get what what you need.

**Reply by ervinpopescu** [reply]
Thanks!

**Comment by m-col** [comment]
This is something that is often brought up WRT documentation, that we essentially omit a huge part of Qtile's configurability via hacking its internals in the config. IIRC a common concern with doing so is how far to document - it's potentially a huge can of worms (with it being Python, anything goes, hack away!) because it becomes quickly difficult to offer help when doing so. That said maybe that's a worry that is unproven so we should probs document it at least a bit. It might be worth mentioning this on https://github.com/qtile/qtile/issues/1679, and likewise other things that aren't documented.

**Reply by elParaguayo** [reply]
Once we merge the decorators PR then this becomes a bit easier as we will have consistency with command names.

**Reply by ervinpopescu** [reply]
> This is something that is often brought up WRT documentation, that we essentially omit a huge part of Qtile's configurability via hacking its internals in the config. IIRC a common concern with doing so is how far to document - it's potentially a huge can of worms (with it being Python, anything goes, hack away!) because it becomes quickly difficult to offer help when doing so. That said maybe that's a worry that is unproven so we should probs document it at least a bit. It might be worth mentioning this on #1679, and likewise other things that aren't documented.

I totally agree this should be documented more :)

**Reply by ervinpopescu** [reply]
> Once we merge the decorators PR then this becomes a bit easier as we will have consistency with command names.

You mean the `cmd_` part? I'm not familiar with this :)

**Reply by elParaguayo** [reply]
Yes. For example, you can do `lazy.spawn` but, to do this in a function you need to call `qtile.cmd_spawn`.

It runs **exactly** the same method but the names are different.
